# Describe the goals of the study.

    With the coronavirus which took the entire world by surprise, bringing about less mobility in human activities.
    City dwellers no longer spent their free time outside, going to cafes and malls; with more people at home, reading
    books. This phenomena attracted the attention of startups that rushed to develop new apps for book lovers.
    Given a database of one of the services competing in this market. This contains data on books, publishers, authors, 
    and customer ratings and reviews of books. This information will be used to generate a value proposition for a 
    new product.

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'practicum_student',         # username
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the database

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# Study the tables (print the first rows)

In [2]:
# print first 5 rows of books table
pd.read_sql_query('''SELECT * FROM books LIMIT 5;''',con=engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# print first 5 rows of authors table
pd.read_sql_query('''SELECT * FROM authors LIMIT 5;''',con=engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# print first 5 rows of publishers table
pd.read_sql_query('''SELECT * FROM publishers LIMIT 5;''',con=engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
# print first 5 rows of ratings table
pd.read_sql_query('''SELECT * FROM ratings LIMIT 5;''',con=engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# print first 5 rows of reviews table
pd.read_sql_query('''SELECT * FROM reviews LIMIT 5;''',con=engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

# Conclusion

    we viewed the first 5 rows of each table in the database.books table comprises of 6 columns, authors table 
    comprises of 2 columns, publisher table comprises of 2 columns, ratings table comprises of 4 columns and finally 
    reviews 4 columns

# Make an SQL query for each of the tasks.

## Find the number of books released after January 1, 2000

In [7]:
pd.read_sql_query('''SELECT 
                       COUNT(book_id) AS number_of_books 
                     FROM books
                     WHERE publication_date::date > '2000-01-01';''',con=engine)

number_of_books
0              819

# Conclusion

     From the given task the number books released after january 1,2000 is 819 in number.

## Find the number of user reviews and the average rating for each book.

In [8]:
pd.read_sql_query('''SELECT 
                       books.title AS title,
                       COUNT(reviews.review_id) AS number_of_reviews,
                       ROUND(AVG(ratings.rating),2) AS average_rating
                     FROM books
                     INNER JOIN reviews ON reviews.book_id=books.book_id
                     INNER JOIN ratings ON ratings.book_id=books.book_id
                     GROUP BY
                       books.title
                     ORDER BY
                       number_of_reviews DESC,
                       average_rating DESC;''',con=engine)

title  number_of_reviews  \
0                              Twilight (Twilight  #1)               1120   
1                  The Hobbit  or There and Back Again                528   
2                               The Catcher in the Rye                516   
3    Harry Potter and the Prisoner of Azkaban (Harr...                492   
4    Harry Potter and the Chamber of Secrets (Harry...                480   
..                                                 ...                ...   
988                     Death: The High Cost of Living                  2   
989                                Moo  Baa  La La La!                  2   
990                                  History of Beauty                  2   
991               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1                  2   
992                 1 000 Places to See Before You Die                  2   

     average_rating  
0              3.66  
1              4.13  
2              3.83  
3              4.41  
4              4.29  
..              ...  
988            3.00  
989            3.00  
990            2.50  
991            2.50  
992            2.50  

[993 rows x 3 columns]

# Conclusion

    There are 993 books in total and out of this the book titled Twilight had the greatest number of reviews at 
    an average rating of 3.66 whiles book titled 1000 places to see before you die has the lowest number of reviews 
    at an average rating of 2.50.

## Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [9]:
pd.read_sql_query('''SELECT 
                       publishers.publisher AS publisher,
                       COUNT(books.book_id) AS number_of_books
                     FROM books
                     INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
                     WHERE books.num_pages > 50
                     GROUP BY
                       publishers.publisher
                     ORDER BY
                       number_of_books DESC;''',con=engine)

publisher  number_of_books
0                                        Penguin Books               42
1                                              Vintage               31
2                             Grand Central Publishing               25
3                                     Penguin Classics               24
4                                     Ballantine Books               19
..                                                 ...              ...
329                                         Turtleback                1
330  Atheneum Books for Young Readers: Richard Jack...                1
331                                     Penguin Signet                1
332                                    Victor Gollancz                1
333                      Harvard Business Review Press                1

[334 rows x 2 columns]

# Conclusion

    Also from this task the publisher with the greatest number of books published are Grand Central Publishing,
    Penguin Books,Black Dog & Leventhal Publishers and Bantam with 3 books published by each publisher.

## Identify the author with the highest average book rating: look only at books with at least 50 ratings.

In [10]:
pd.read_sql_query('''SELECT 
                       authors.author AS author,
                       ROUND(AVG(ratings.rating),3) AS average_rating,
                       COUNT(ratings.rating) AS number_of_rating
                     FROM authors
                     INNER JOIN books ON books.author_id=authors.author_id
                     INNER JOIN ratings ON ratings.book_id=books.book_id
                     GROUP BY
                       authors.author
                     HAVING 
                       COUNT(ratings.rating) >= 50
                     ORDER BY
                        number_of_rating DESC;''',con=engine)



author  average_rating  \
0                          J.K. Rowling/Mary GrandPré           4.288   
1                                      J.R.R. Tolkien           4.241   
2                                     Stephenie Meyer           3.663   
3                                           Dan Brown           3.741   
4                                     Nicholas Sparks           3.883   
5                                        Stephen King           4.009   
6                                       J.D. Salinger           3.847   
7                                        Rick Riordan           4.131   
8                                        Jodi Picoult           3.882   
9             George Orwell/Boris Grabnar/Peter Škerl           3.730   
10                                     John Steinbeck           3.644   
11                                    William Golding           3.901   
12                                       John Grisham           3.971   
13  William Shakespeare/Paul Werstine/Barbara A. M...           3.788   
14                                         Lois Lowry           3.738   
15                                    James Patterson           3.859   
16                           Roald Dahl/Quentin Blake           4.210   
17                                    Sophie Kinsella           3.877   
18           Paulo Coelho/Alan R. Clarke/Özdemir İnce           3.789   
19                                      Arthur Golden           4.107   
20                                  Louisa May Alcott           4.204   
21                  Markus Zusak/Cao Xuân Việt Khương           4.264   
22                                    Agatha Christie           4.283   
23                                     Diana Gabaldon           4.300   

    number_of_rating  
0                312  
1                166  
2                160  
3                143  
4                111  
5                106  
6                 98  
7                 84  
8                 76  
9                 74  
10                73  
11                71  
12                70  
13                66  
14                65  
15                64  
16                62  
17                57  
18                57  
19                56  
20                54  
21                53  
22                53  
23                50

# Conclusion

    The author with the highest average book rating happens to be Diana Gabaldon with an average rating of 4.3 
    with 50 number of ratings.

# Find the average number of text reviews among users who rated more than 50 books

In [11]:
pd.read_sql_query('''SELECT 
                        round(AVG(sub.number_of_reviews),2) AS average_reviews
                     FROM 
                        (SELECT
                            username,
                            COUNT(reviews.text) AS number_of_reviews
                        FROM reviews
                        WHERE 
                           username IN(
                                SELECT 
                                   username
                                FROM ratings
                                GROUP BY
                                  username
                                HAVING
                                   COUNT(ratings.book_id) > 50 
                                )
                        GROUP BY
                           username) AS sub;''',con=engine)



average_reviews
0            24.33

# Conclusion

    The average number of text reviews among users with more than 50 books is 24.33